File: C/aleph.c
/* ALEPH - A small-state CSPRNG and Stream Cipher
   ALEPH is a ALEPH with an 8+5-word internal state
   ALEPH may be seeded with a 256-bit key
   ALEPH Copyright C.C.Kayne 2014, GNU GPL V.3, cckayne@gmail.com
   ALEPH is inspired by Bob Jenkins' PRNG insights (Public Domain).
*/
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include "aleph.h"

#define TEST
#define VERBOSE

// ALEPH defines

//#define ALEPH8
#define ALEPH16
//#define ALEPH32
//#define ALEPH64

/* internal state parameters */
#ifdef ALEPH8
#define STSZ 8
#define NAME "ALEPH8"
#endif
#ifdef ALEPH16
#define STSZ 16
#define NAME "ALEPH16"
#endif
#ifdef ALEPH32
#define STSZ 32
#define NAME "ALEPH32"
#endif
#ifdef ALEPH64
#define STSZ 64
#define NAME "ALEPH64"
#endif
#define STM1 STSZ-1
#define STBYTES STSZ*4
#define STBITS 128+STBYTES*8

/* 2**32/phi, where phi is the golden ratio */
#define GOLDEN   0x9e3779b9
/* indirection counter */
#define ALEPH (++a)

static u4 state[STSZ], rsl[STSZ], a, b, c, d, e, rcnt=0;

#ifdef TEST
static void statepeek(void);
#endif


static void aleph(void) {
	u4 i;
	for (i=0; i<STSZ; i++) {
		e = state[d & STM1] + ALEPH;
		state[d & STM1] = b ^ c;
		b = c - d;
		c = d + e;
		rsl[i] = d = e + state[i];
	}
	#ifdef TEST
	statepeek();
	#endif
}


// obtain a ALEPH pseudo-random value in [0..2**32]
u4 aleph_Random(void) {
	u4 r = rsl[rcnt];
	++rcnt;
	if (rcnt==STSZ) {
		aleph();
		rcnt = 0;
	}
	return r;
}


void aleph_Reset(void) {
	register u4 i,r;
	rcnt = 0;
	a = b = c = d = e = GOLDEN;
	for (i=0; i<STSZ; i++) { state[i]=GOLDEN; rsl[i]=GOLDEN; }
}


// seed ALEPH with a 2048-bit block of 4-byte words (Bob Jenkins method) 
void aleph_SeedW(char *seed, int rounds)
{
	register u4 i,l;
	char s[STBYTES*2];
	l=strlen(seed);
	if (l>STBYTES) l=STBYTES;
	memset(s,0,l+1);
	/* truncate seed to state-size if necessary */
	for (i=0; i<l; i++) s[i] = seed[i];
	aleph_Reset();
	memcpy((char *)state, (char *)s, l);
	/* fatten the variables on some seed-bytes */
	a+=state[3]; b+=state[2]; c+=state[1]; d+=state[0]; 
	aleph();
	for (i=0; i<rounds; i++) aleph_Random();  
}


// ALEPH # of bits internal state
u4 aleph_StateBits(void) {
	return STBITS;
}


// ALEPH expected cycle length
u4 aleph_Cycle(void) {
	return (STBITS+1)/2;
}


// ALEPH maximum key length (bits)
u4 aleph_KeyLength(void) {
	return STBYTES*8;
}


// ALEPH Name
char* aleph_Name(void) {
	return NAME;
}


#ifdef TEST
static u4 bcnt=0;
void testinit(u4 val) {
	register u4 i;
	rcnt = 0;
	b = c = d = e = val;
	for (i=0; i<STSZ; i++) { state[i]=val; rsl[i]=0; }
}
static void statepeek(void) {
	register u4 i;
	++bcnt;
	printf("%3u) aleph%u...\n",bcnt,STSZ);
	for (i=0; i<STSZ; i++) {
		#ifdef VERBOSE
		printf("state %3u: %11u %c %02X  | rsl %3u: %11u %c %02X\n",
			i,state[i],state[i] % 26 + 'A',state[i] & 255,
			i,rsl[i],rsl[i] % 26 + 'A',rsl[i] & 255);
		#endif
	}
	printf("       a = %11u %c %02X\n       b = %11u %c %02X\n       c = %11u %c %02X\n       d = %11u %c %02X\n       e = %11u %c %02X\n",
		a,a % 26+'A',a & 255,b,b % 26+'A',b & 255,c,c % 26+'A',c & 255,d,d % 26+'A',d & 255,e,e % 26+'A',e & 255);
}
#endif

File: Pascal/aleph.pas
unit aleph;
(* ALEPH - A minimal CSPRNG reflecting the Zen of cipher design
   ALEPH is an ALEPH with a 16,32 or 64+4-word internal state
   ALEPH may be seeded with a 512-, 1024- or 2048-bit key
   ALEPH Copyright C.C.Kayne 2014, GNU GPL V.3, cckayne@gmail.com
   ALEPH is inspired by Bob Jenkins' PRNG insights (Public Domain).
*)
{$mode delphi}
{$inline on}
//are we testing?
{$define TEST}
//verbose test output
{$define VERBOSE}

interface

// initialize/reset ALEPH
procedure alReset;
// obtain an ALEPH pseudo-random value in [0..2**32]
function alRandom: Cardinal;
// seed ALEPH with a 2048-bit block of 4-byte words (Bob Jenkins method) 
procedure alSeedW(seed: string; rounds: integer);
// ALEPH # of bits internal state
function alStateBits: Cardinal;
// ALEPH expected cycle length
function alCycle: Cardinal;
// ALEPH Name
function alName: string;

implementation
{$ifdef TEST}uses StrUtils;{$endif}

// select your ALEPH variant here
{$define ALEPH16}
{ $define ALEPH32}
{ $define ALEPH64}

const 	
		{$ifdef ALEPH16}
		NAME = 'ALEPH16';
		STSZ = 16;
		{$endif}
		{$ifdef ALEPH32}
		NAME = 'ALEPH32';
		STSZ = 32;
		{$endif}
		{$ifdef ALEPH64}
		NAME = 'ALEPH64';
		STSZ = 64;
		{$endif}
		
		STM1 = STSZ-1;
		STBYTES = STSZ*4;
		STBITS = 160+STBYTES*8;
		// 2**32/phi, where phi is the golden ratio
		GOLDEN   = $9e3779b9;

		MODU = 26;
		START= 65;
		
// ALEPH STATE
var		rsl: 	array[0..STSZ] of Cardinal;
		state: 	array[0..STSZ] of Cardinal;
		a,b,c,d,e,rcnt: Cardinal;

	
{$ifdef TEST}
var bcnt: Cardinal = 0;
procedure statepeek; 
	var i: Cardinal;
	begin
		inc(bcnt);
		Writeln(bcnt:3,') aleph ...');
		for i:=0 to STM1 do begin
			{$ifdef VERBOSE}
			Writeln('state ',
				i:3,': ',state[i]:11,chr(state[i] mod MODU + START):2,
					dec2numb((state[i] and 255),2,16):3,'  | rsl ',
				i:3,': ',rsl[i]:11,chr(rsl[i] mod MODU + START):2,
					dec2numb((rsl[i] and 255),2,16):3);
			{$endif}
		end;
		Writeln('       a = ',a:11,chr(a mod MODU+START):2,dec2numb((a and 255),2,16):3);     
		Writeln('       b = ',b:11,chr(b mod MODU+START):2,dec2numb((b and 255),2,16):3);     
		Writeln('       c = ',c:11,chr(c mod MODU+START):2,dec2numb((c and 255),2,16):3);    
		Writeln('       d = ',d:11,chr(d mod MODU+START):2,dec2numb((d and 255),2,16):3);
		Writeln('       e = ',e:11,chr(e mod MODU+START):2,dec2numb((e and 255),2,16):3);
	end;
{$endif}


// ALEPH is filled every 16, 32 or 64 rounds
procedure aleph;
    var i: Cardinal;
	begin
		for i:=0 to STM1 do begin
			inc(a);
			e := state[d and STM1] + a;
			state[d and STM1] := b xor c;
			b := c - d;
			c := d + e;
			d := e + state[i];
			rsl[i] := d;			
		end;
		{$ifdef TEST}
		statepeek;
		{$endif}
	end;
	
	
// initialize/reset ALEPH (obligatory)
procedure alReset;
	var i: Cardinal;
	begin
		rcnt:=0;
		a := GOLDEN; b := a; c := b; d:=c; e:=d;
		for i:=0 to STM1 do begin
			state[i] := GOLDEN; rsl[i] := 0;
		end;
	end;


// obtain a ALEPH pseudo-random value in [0..2**32]
function alRandom: Cardinal;
	begin
		alRandom := rsl[rcnt];
		inc(rcnt);
		if (rcnt=STSZ) then begin
			aleph;
			rcnt := 0;
		end;
	end;


// seed ALEPH with a 2048-bit block of 4-byte words (Bob Jenkins method) 
procedure alSeedW(seed: string; rounds: integer);
	var i,l: Cardinal;
		p: pointer;
	begin
		p:=@state[0];
		l:=Length(seed);
		if (l>STBYTES) then l:=STBYTES;
		alReset;
		for i:=0 to l-1 do
			byte((p+i)^) := byte(seed[i+1]);
		// fatten the variables on some seed-bytes
		a+=state[3]; b+=state[2]; c+=state[1]; d+=state[0]; 
		aleph;
		for i:=1 to rounds do alRandom;  
	end;


// ALEPH # of bits internal state
function alStateBits: Cardinal;
	begin
		alStateBits := STBITS;
	end;


// ALEPH expected cycle length
function alCycle: Cardinal;
	begin
		alCycle := (STBITS+1) div 2;
	end;


// ALEPH Name
function alName: string;
	begin
		alName := NAME;
	end;

end.